File: 120_triangle/minimumTotal.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype int
        """
        T = []
        for i in range(len(triangle)):
            T.append([0]*(i+1))
            
        T[0][0] = triangle[0][0]
        
        for layer in range(1, len(triangle)):
            T[layer][0] = T[layer - 1][0] + triangle[layer][0]
            T[layer][-1]= T[layer - 1][-1] + triangle[layer][-1]
            for idx in range(1, len(triangle[layer]) - 1):
                T[layer][idx] = min(T[layer - 1][idx], T[layer - 1][idx - 1]) + triangle[layer][idx]
        print T 
        return min(T[-1])
    
if __name__ == '__main__':
    sol = Solution()
    triangle = [[2], [3,4], [6,5,7], [4,1,8,3]]
    print sol.minimumTotal(triangle)

File: 120_triangle/minimumTotal.cpp
#include <vector>

using namespace std;

class Solution {
    public:
        int minimumTotal(vector<vector<int> >& triangle) {
            int n = triangle.size();
            vector<vector<int> > dp(n, vector<int>(n, 0));
            int minSum = INT_MAX;
            int i, j;

            // base case
            dp[0][0] = triangle[0][0];

            // the left side
            for(i = 1; i < n; i++)
                dp[i][0] = triangle[i][0] + dp[i - 1][0];

            // the right side
            for(i = 1; i < n; i++) {
                int len = triangle[i].size();
                dp[i][len - 1] = triangle[i][len - 1] + dp[i - 1][triangle[i-1].size() - 1]; 
            }

            // general case
            for(i = 1; i < n; i++) {
                for(j = 1; j < triangle[i].size() - 1; j++) {
                    dp[i][j] = min(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i][j];
                }
            }

            for(i = 0; i < triangle[n - 1].size(); i++) {
                minSum = min(minSum, dp[n - 1][i]);
            }

            return minSum;
        }
};

int main() {
    return 0;
}
